feat(Topology/Order): topological basis of scott topology on Complete…

edwin1729 opened the pull request:
…PartialOrder


---
[Zulip discussion](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/.3E.20Algebraic.20complete.20partial.20order.20and.20domain.20theory/with/536359313) for this PR.
(2/5) PRs in domain theory, proving that scott topologies over Algebraic DCPOs (`CompletePartialOrder`) are sober.
The main reference is [Renata, Duality in Domain Theory](https://alyata.github.io/documents/duality_domain_theory.pdf). But the statements can also be found in the canonical text [Abramsky and Jung](https://www.cs.ox.ac.uk/people/samson.abramsky/handbook.pdf)
 This first PR proves two prerequisites, namely:
- the specialization order induced by the scott topology corresponds to the existing order of the DCPO. Prop 3.1.5 in [Duality in Domain Theory](https://alyata.github.io/documents/duality_domain_theory.pdf) and Prop 2.3.2(1) in [Abramsky and Jung](https://www.cs.ox.ac.uk/people/samson.abramsky/handbook.pdf)
- the upward closures of compact elements of the DCPO form a topological basis for the Scott Topology. Prop 3.5.2 in [Duality in Domain Theory](https://alyata.github.io/documents/duality_domain_theory.pdf) and Prop 2.3.6(2) in [Abramsky and Jung](https://www.cs.ox.ac.uk/people/samson.abramsky/handbook.pdf)

- [x] depends on: #33061 

**The next PR is here:** #31670 


<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

When merging, all the commits will be squashed into a single commit
listing all co-authors.

Co-authors in the squash commit are gathered from two sources:

First, all authors of commits to this PR branch are included. Thus,
one way to add co-authors is to include at least one commit authored by
each co-author among the commits in the pull request. If necessary, you
may create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

Second, co-authors can also be listed in lines at the very bottom of
the commit message (that is, directly before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines
at the bottom of the commit message (before the `---`, and also before
any "Co-authored-by" lines) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

edwin1729 committed:
feat(Topology/Order): topological basis of scott topology on CompletePartialOrder

edwin1729 commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,292 @@
+  rw [specializes_iff_forall_open]
+  constructor
+  · intro x_le_y u hu x_in_u
+    apply (@isUpperSet_of_isOpen α {d | DirectedOn (· ≤ ·) d }) at hu
```
This lemma is not introduced by me, and I'm not sure how to get better inference

plp127 commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,292 @@
+    The weaker version is still useful as it is easier to use when sufficient.
+    I don't reuse the previous result to prove this, since the proof turns out just as long -/
+lemma open_eq_open_of_basis' (u : Opens D) :
+  u = sSup ({ o | ∃ (c: D) (hc: c ∈ 𝕂 D), c ∈ u ∧ (o = ⟨c,hc⟩ᵘᵒ) }) := by
```
Oh, did you not know about the notation? You can write it as `⨆ (c : D) (hc : c ∈ 𝕂 D) (hu : c ∈ u), ⟨c, hc⟩ᵘᵒ`.

edwin1729 commented:
@plp127 I didn't realize CI was failing. This is ready for review again now.

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,284 @@
+/-!
+# Scott Complete Partial Order
+
+Properties of Scott topologies over (Directed) `CompletePartialOrder`, and Algberaic DCPOs.
```
```suggestion
Properties of Scott topologies over (Directed) `CompletePartialOrder`, and Algebraic DCPOs.
```

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,284 @@
+- `Compact` - The element x is compact if for any directed subset A ⊆ D, whenever `x ≤ sSup A`,
+there is some a ∈ A s.t. x ≤ a already.
+- `AlgebraicDCPO` - an extension of `CompletePartialOrder` which is additionally "algebraic".
+Algebraicity has a correspondence with axioms of a Scott topology
```
Can you be more specific about what "has a correspondence" means here?

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,284 @@
+- `isTopologicalBasis_of_Ici_compact`: The upward closures of compact elements form a topological
+basis under the Scott Topology. Prop 3.5.2 in [reneta2025]
+- `open_eq_of_open_basis'`: Explicit construction an Open set from the basis mentioned above.
+Prop 2.3.6(2) [abramsky_gabbay_maibaum_1994]
```
```suggestion
- `specialization_iff_ge`: The original order and the specialization order induced by the Scott
  topology, correspond. Prop 3.5.2 in [reneta2025]. Prop 2.3.2(1) in [abramsky_gabbay_maibaum_1994]
- `isTopologicalBasis_of_Ici_compact`: The upward closures of compact elements form a topological
  basis under the Scott Topology. Prop 3.5.2 in [reneta2025]
- `open_eq_of_open_basis'`: Explicit construction an Open set from the basis mentioned above.
  Prop 2.3.6(2) [abramsky_gabbay_maibaum_1994]
```
and similar below

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,284 @@
+/-- Encodes notion of observable properties in programs (points are program semantics) -/
+class AlgebraicDCPO (α : Type*) extends CompletePartialOrder α where
+  algebraic : ∀ x : α, (compactLowerSet x).Nonempty ∧ DirectedOn (· ≤ ·) (compactLowerSet x) ∧
+  x = sSup (compactLowerSet x)
```
```suggestion
    x = sSup (compactLowerSet x)
```

b-mehta reviewed (commented):
Here are some preliminary comments; we discussed more in person.

edwin1729 commented:
- [ ] depends on: #33061 

The above PR has been introduced to generalise the definition of `Compact` (`IsCompactElement` in Mathlib),
to avoid redefining this notion here. As discussed with @b-mehta

edwin1729 committed:
refactor: introduce an sSup based open_of_open_eq_basis

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,295 @@
+  [IsScott α {d | DirectedOn (· ≤ ·) d}]
+
+/-- The order from `CompletePartialOrder` and the specialization order induced by the Scott
+topology, correspond. Unfortunately Mathlib's specialization order `⤳` is opposite to `≤`
```
```suggestion
topology, correspond. Unfortunately Mathlib's specialization order `⤳` is opposite to `≤`.
```

b-mehta reviewed (commented):
I've left a few comments. 
In general, public definitions in mathlib are somewhat expensive, so ensure that you only make definitions you actually need. 
Also, I think lots of the docstrings could be improved for mathlib.

edwin1729 committed:
improved comments and removing `CompactSet` defs based refactors

edwin1729 commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,295 @@
+  topology, correspond. Prop 3.5.2 in [reneta2024]. Prop 2.3.2(1) in [abramsky_gabbay_maibaum_1994]
+- `isTopologicalBasis_of_Ici_compact`: The upward closures of compact elements form a topological
+  basis under the Scott Topology. Prop 3.5.2 in [reneta2024]
+- `open_eq_of_open_basis'`: Explicit construction an Open set from the basis mentioned above.
```
Sorry, it was renamed slightly, it is now fixed.

edwin1729 committed:
move some lemmas to the next PR, since it's hard to review without further context

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,295 @@
+/-- Any open set, `u`, can be constructed as a union of sets from the basis.
+    The basis consists of the upward closures of those compact elements in `u`
+    This is the weaker version of the lemma using `Set`s instead of `Opens`. -/
+lemma open_eq_open_of_basis (u : Set D) (hu : IsOpen u) :
```
Is [TopologicalSpace.IsTopologicalBasis.open_eq_sUnion'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Bases.html#TopologicalSpace.IsTopologicalBasis.open_eq_sUnion') or [TopologicalSpace.IsTopologicalBasis.open_iff_eq_sUnion](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Bases.html#TopologicalSpace.IsTopologicalBasis.open_iff_eq_sUnion) closer to what you want then? In either case, can be discussed for the next PR :)

(In general, if you can make PRs smaller, that would make them easier to review, especially for things like this!)

edwin1729 committed:
fix typo in reference

edwin1729 committed:
remove extra unused abbrevs for compact elements

edwin1729 committed:
replace `DirectedOn` sets with univ as former was reduntant

edwin1729 committed:
use Opens.IsBasis instead of TopologicalSpace.IsTopologicalBasis

This is much more useful for the later goals, particularly since we are
working with a existing mathlib code which uses `Opens` not `Set`

edwin1729 commented:
@b-mehta could you please take another look? I believe this PR is complete.

b-mehta committed:
Merge branch 'master' into scott-topology

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,226 @@
+element and 2) every element is given by the supremum of a set of compact elements (algebraic). -/
+class AlgebraicDCPO (α : Type*) extends CompletePartialOrder α, OrderBot α where
+  algebraic : ∀ x : α, ({y : α | IsCompactElement y ∧ y ≤ x}).Nonempty ∧ DirectedOn (· ≤ ·)
+    {y : α | IsCompactElement y ∧ y ≤ x} ∧ x = sSup {y : α | IsCompactElement y ∧ y ≤ x}
```
Since we have a bot, the `Nonempty` is only saying something in the case of `x != bot`, and in fact in that case it just says that bot is compact.

edwin1729 replied:
Yes it was redundant, and removed. `IsCompactElement_bot` is extracted as a lemma as it is used again in a later PRs

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,226 @@
+
+/-- An algebraic directed complete partial order is a `CompletePartialOrder` with 1) a least
+element and 2) every element is given by the supremum of a set of compact elements (algebraic). -/
+class AlgebraicDCPO (α : Type*) extends CompletePartialOrder α, OrderBot α where
```
CompletePartialOrder now implies OrderBot, so you don't need to add this

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,226 @@
+/-- The order from `CompletePartialOrder` and the specialization order induced by the Scott
+topology, correspond. Unfortunately Mathlib's specialization order `⤳` is opposite to `≤`.
+Prop 3.5.2 in [renata2024]. Prop 2.3.2(1) in [abramsky_gabbay_maibaum_1994]. -/
+lemma specialization_iff_ge {x y : α} : x ≤ y ↔ y ⤳ x := by
```
The statement as you've written it is "ge_iff_specialization", but I think the right fix here is to swap the statement so it's the other way round (and then the name is correct)

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,226 @@
+    simp only [Ici, mem_setOf_eq] at hx ⊢
+    transitivity x
+    · exact hx
+    · exact x_le_y
```
```suggestion
    grind [IsUpperSet]
```
I think some of the other proofs can be shorter like this too

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,226 @@
+
+/-- The upward closure of a compact element (`Ici e`) is an open set.
+We refer to the `Ici e` as basis due to `isTopologicalBasis_Ici_image_compactSet`. -/
+lemma isOpen_of_basis (e : α) (he₀ : IsCompactElement e) : IsOpen (Ici e) := by
```
```suggestion
lemma IsCompactElement.isOpen_Ici (e : α) (he₀ : IsCompactElement e) : IsOpen (Ici e) := by
```
I understand the original name, but the naming convention should still apply here

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,226 @@
+/-- Given any point `x` in `D` in an open set `u`, there exists
+an upward closure of a compact element (`Ici e`), within `u` which contains `x`.
+We refer to the `Ici e` as basis due to `isTopologicalBasis_Ici_image_compactSet`. -/
+lemma exists_basis_mem_basis (x : D) (u : Set D) (x_in_u : x ∈ u) (hu : IsOpen u)
```
as before, let's stick to `Ici` instead of `basis`

edwin1729 replied:
I really don't know what to name this lemma. I have proposed `exists_Ici_mem` (supposed to mean exists an Ici such that Ici is a mem of `u`, the input)

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,226 @@
+    simp only [image_univ, mem_image, mem_range, Subtype.exists, ↓existsAndEq, coe_mk, true_and,
+      exists_prop, exists_exists_and_eq_and, mem_Ici]
+    use c, hc
+    exact ⟨mem_Ici.1 x_in_c', Subset.refl (LE.le (Ici c)) hc'⟩
```
```suggestion
theorem isTopologicalBasis_Ici_image_compactSet :
    IsTopologicalBasis {Ici c | (c : D) (_ : IsCompactElement c)} := by
  apply isTopologicalBasis_of_isOpen_of_nhds
  · grind [isOpen_of_basis]
  · grind [exists_basis_mem_basis]
```

edwin1729 replied:
I have chosen to use `Opens.IsBasis` over IsTopologicalBasis, since its generality is required in later parts of the proofs. Particularly `IsBasis` talks about `Set (Opens D)` vs `IsTopologicalBasis` talking about `Set (Set D)`. Briefly, this by the existing definition in mathlib of `[Locale.localePointOfSpacePointof](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Order/Category/FrameAdjunction.html#Locale.localePointOfSpacePoint)` (it uses `Opens D` not `Set D`).

b-mehta replied:
What I'd really like is to avoid needing the `abbrev`s above though - `abbrev`s are fine for a small project, but these ones aren't really adding much to mathlib. 
Let me also note that your proof starts off by applying `isTopologicalBasis_of_isOpen_of_nhds`, which is about `IsTopologicalBasis` (ie `IsTopologicalBasis` is actually more general, which is why it applies to `IsBasis`). It's also easy to convert a `Set (Set D)` into `Set (Opens D)` using Set.preimage (and its notation)

Commit:
feat(Topology/Order): use grind more, remove redundancy in `AlgebraicDCPO` def

edwin1729 commented:
-awaiting-author

b-mehta commented:
Please avoid force-pushing, since it makes it really hard to re-review!

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,216 @@
+      ∀ s : Set α, s.Nonempty → DirectedOn (· ≤ ·) s → k ≤ sSup s → ∃ x : α, x ∈ s ∧ k ≤ x := by
+  constructor
+  · intro hk s hs hs' h_le
+    exact hk s (sSup s) hs hs' (lubOfDirected s hs') h_le
```
```suggestion
    exact hk s (sSup s) hs hs' hs'.isLUB_sSup h_le
```

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,216 @@
+  · intro hk s hs hs' h_le
+    exact hk s (sSup s) hs hs' (lubOfDirected s hs') h_le
+  · intro h s u hs hs' hu h_le
+    have u_eq_sSup: u = sSup s := IsLUB.unique hu (lubOfDirected s hs')
```
```suggestion
    have u_eq_sSup : u = sSup s := hu.unique hs'.isLUB_sSup
```

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,216 @@
+  rw [isCompactElement_iff_le_of_directed_sSup_le]
+  intro s ⟨e, he⟩ _ _
+  use e, he
+  exact OrderBot.bot_le e
```
```suggestion
  intro s _ ⟨e, he⟩ _ _ _
  exact ⟨e, he, bot_le⟩
```

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,216 @@
+    exact h s hs hs' h_le
+
+/-- `⊥` in a `CompletePartialOrder` is compact. -/
+lemma isCompactElement_bot : IsCompactElement (⊥ : α) := by
```
This lemma and the previous should probably go in the CompletePartialOrder file, rather than being here.

edwin1729 replied:
Do you mean `Mathlib/Order/CompactlyGenerated/Basic.lean`? I have moved it there.

b-mehta replied:
Unfortunately I did mean Mathlib/Order/CompletePartialOrder; I think putting it in CompactlyGenerated/ is why there's now a large-import. But maybe putting it in CompletePartialOrder will give a large-import too. If that's the case, then you can make a new file Mathlib/Order/CompactlyGenerated/CompletePartialOrder.lean for these lemmas. Actually now that I look at it, `isCompactElement_bot` doesn't require it to be a complete partial order at all, right? That one's true just in a `PartialOrder` that satisfies `OrderBot`

b-mehta replied:
Ouch, looks like it does give a large-import too. So the new file is a good idea, but note that `isCompactElement_bot` should still go in CompactlyGenerated/ because it doesn't need to know about CompletePartialOrder

edwin1729 replied:
Well I just noticed that `isCompactElement_iff_le_of_directed_sSup_le` is never used, since it has been swept under grind presumably now. So the new file for this is no longer required?

b-mehta replied:
Oh good point, maybe we can just drop the lemma?

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,216 @@
+element and 2) every element is given by the supremum of a set of compact elements (algebraic). -/
+class AlgebraicDCPO (α : Type*) extends CompletePartialOrder α where
+  algebraic (x : α) : DirectedOn (· ≤ ·) {y : α | IsCompactElement y ∧ y ≤ x} ∧
+    x = sSup {y : α | IsCompactElement y ∧ y ≤ x}
```
instead of having the `∧`, it would make more sense to have these as separate fields. for example:
```
  directedOn_isCompactElement_le (x : α) : DirectedOn (· ≤ ·) {y : α | IsCompactElement y ∧ y ≤ x}
  isLUB_isCompactElement_le (x : α) : IsLUB {y : α | IsCompactElement y ∧ y ≤ x} x
```

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,216 @@
+    exact h_specialize
+  · intro x_le_y u hu x_in_u
+    apply isUpperSet_of_isOpen univ at hu
+    exact hu x_le_y x_in_u
```
```suggestion
lemma specialization_iff_ge {x y : α} : y ⤳ x ↔ x ≤ y := by
  rw [specializes_iff_forall_closed]
  constructor
  · intro h
    simpa using h (Iic y) isClosed_Iic (by simp)
  · intro hxy s hs hys
    exact isLowerSet_of_isClosed hs hxy hys
```

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,216 @@
+      simp only [Ici, mem_setOf_eq] at hx'
+      exact hx'
+    choose a a_in_d ha' using he₀ d nonempty hd he₁
+    exact ⟨a, a_in_d, mem_Ici.mpr ha'⟩
```
```suggestion
lemma IsCompactElement.isOpen_Ici (e : α) (he₀ : IsCompactElement e) : IsOpen (Ici e) := by
  rw [isOpen_iff_isUpperSet_and_dirSupInaccOn univ]
  constructor
  · grind [IsUpperSet]
  · rw [dirSupInaccOn_univ]
    intro d nonempty hd x hx hx'
    exact he₀ d x nonempty hd hx hx'
```

edwin1729 committed:
Update Mathlib/Topology/Order/ScottTopologyDCPO.lean

Co-authored-by: Bhavik Mehta <bhavikmehta8@gmail.com>

edwin1729 committed:
Update Mathlib/Topology/Order/ScottTopologyDCPO.lean

Co-authored-by: Bhavik Mehta <bhavikmehta8@gmail.com>

edwin1729 committed:
Update Mathlib/Topology/Order/ScottTopologyDCPO.lean

Co-authored-by: Bhavik Mehta <bhavikmehta8@gmail.com>

edwin1729 committed:
Update Mathlib/Topology/Order/ScottTopologyDCPO.lean

Co-authored-by: Bhavik Mehta <bhavikmehta8@gmail.com>

edwin1729 committed:
Update Mathlib/Topology/Order/ScottTopologyDCPO.lean

Co-authored-by: Bhavik Mehta <bhavikmehta8@gmail.com>

Commit:
feat(Topology/Order): add fields to AlgebraicDCPO, move CompletePartialOrder lemmas

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,146 @@
+section CompletePartialOrder
+
+variable {α : Type*} [TopologicalSpace α] [CompletePartialOrder α]
+  [IsScott α univ]
```
```suggestion
variable {α : Type*} [TopologicalSpace α] [CompletePartialOrder α] [IsScott α univ]
```

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,146 @@
+class AlgebraicDCPO (α : Type*) extends CompletePartialOrder α where
+  directedOn_isCompactElement_le (x : α) : DirectedOn (· ≤ ·) {y : α | IsCompactElement y ∧ y ≤ x}
+  isLUB_isCompactElement_le (x : α) : IsLUB {y : α | IsCompactElement y ∧ y ≤ x} x
+
```
If you add
```
export AlgebraicDCPO (directedOn_isCompactElement_le isLUB_isCompactElement_le)
```
here, then you can write `directedOn_isCompactElement_le` below instead of `AlgebraicDCPO.directedOn_isCompactElement_le`.

b-mehta commented on Mathlib/Topology/Order/ScottTopologyDCPO.lean:
```diff
@@ -0,0 +1,146 @@
+/-- Given any point `x` in `D` in an open set `u`, there exists
+an upward closure of a compact element (`Ici e`), within `u` which contains `x`. -/
+lemma exists_Ici_isCompactElement_Ici_mem (x : D) (u : Set D) (x_in_u : x ∈ u) (hu : IsOpen u)
+    : ∃ c, IsCompactElement c ∧ x ∈ Ici c ∧ Ici c ⊆ u := by
```
I think it's nice to first prove
```
lemma DirSupInacc.exists_le_isCompactElement
    (x : D) (u : Set D) (x_in_u : x ∈ u) (hu : DirSupInacc u) :
    ∃ c ≤ x, IsCompactElement c ∧ c ∈ u := by
```
which doesn't require that D has the scott topology (only that it's an algebraic dcpo), and then this theorem is a nice combination of scott + algebraic

b-mehta replied:
```
lemma DirSupInacc.exists_le_isCompactElement
    (x : D) (u : Set D) (x_in_u : x ∈ u) (hu : DirSupInacc u) :
    ∃ c ≤ x, IsCompactElement c ∧ c ∈ u := by
  have := hu ⟨⊥, by simp⟩ (directedOn_isCompactElement_le x) (isLUB_isCompactElement_le x) x_in_u
  simp [inter_nonempty] at this
  grind
```
(which will require you to make `isCompactElement_bot` a simp lemma)

edwin1729 committed:
Merge branch 'master' into scott-topology

edwin1729 committed:
move lemmas from CompactlyGenerated.Basic to CompletePartialOrder

edwin1729 committed:
move `isCompactElement_bot` to CompactlyGenerated, delete `isCompactElement_iff_le_of_directed_sSup_le`

edwin1729 committed:
feat: switch to `DirSupInaccOn.exists_le_isCompactElement`
